### PR TITLE
[RN] Prevent multiplying remote thumbnails

### PR DESCRIPTION
**Pull request by lyubomir**

### CODE REVIEW

File: react/features/base/redux/StateListenerRegistry.js [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+                this._listener.bind(
+                    this,
+                    {
+                        /**
+                         * The previous selections of the {@code Selector}s
+                         * registered with this {@code StateListenerRegistry}.
+                         *
+                         * @type Map<any>
+                         */
+                        prevSelections: new Map(),
```
Comment (lyubomir) [inline react/features/base/redux/StateListenerRegistry.js]: Obviously, I didn't change the uses of `prevSelections` to use a `Map` and they rather add properties to it. But that's immaterial to the abstraction right now and I'll fix it later on should be decide to go with it.

File: react/features/base/redux/StateListenerRegistry.js [hunk 2]
```diff
@@ -0,0 +1,136 @@
+// @flow
+
+const logger = require('jitsi-meet-logger').getLogger(__filename);
+
+/**
+ * The type listener supported for registration with
+ * {@link StateListenerRegistry} in association with a {@link Selector}.
+ */
+type Listener = (selection: any, store: Store, prevSelection: any) => void;
```
Comment (paweldomas) [inline react/features/base/redux/StateListenerRegistry.js]: can we have the arguments explained ?

File: react/features/base/participants/reducer.js [hunk 3]
```diff
@@ -147,31 +142,67 @@ function _participant(state: Object = {}, action) {
...
+        // XXX The local participant is not identified in association with a
+        // JitsiConference because it is identified by the very fact that it is
+        // the local participant.
+        conference = undefined;
 
-    default:
-        return state;
+        // id
+        //
+        // XXX The situation of not having an ID for a remote participant should
```
Comment (saghul) [inline react/features/base/participants/reducer.js]: I know you're just moving code around, but since we are here, let's fix it: this is the local participant, so I think the comment can go away.

File: react/features/base/redux/StateListenerRegistry.js [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+     * the specified redux store may potentially have changed).
+     *
+     * @param {Store} store - The redux store to which this
+     * {@code StateListenerRegistry} is to {@code subscribe}.
+     * @returns {void}
+     */
+    subscribe(store: Store) {
+        // XXX If StateListenerRegistry is not utilized by the app to listen to
+        // state changes, do not bother subscribing to the store at all.
+        if (this._selectorListeners.size) {
```
Comment (saghul) [inline react/features/base/redux/StateListenerRegistry.js]: I wonder if it wouldn't be safer to do this on the first register instead?

File: react/features/base/redux/StateListenerRegistry.js [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+        for (const selectorListener of this._selectorListeners) {
+            const prevSelection = prevSelections.get(selectorListener);
+
+            try {
+                const selection
+                    = selectorListener.selector(
+                        store.getState(),
+                        prevSelection);
+
+                if (prevSelection !== selection) {
```
Comment (saghul) [inline react/features/base/redux/StateListenerRegistry.js]: I don't think this is enough for anything other than simple objects. How about `_.isEqual` instead?

File: react/features/base/participants/middleware.js [hunk 6]
```diff
@@ -127,6 +128,24 @@ MiddlewareRegistry.register(store => next => action => {
     return next(action);
 });
 
+// Syncs the redux state features/base/participants up with the redux state
```
Comment (saghul) [inline react/features/base/participants/middleware.js]: style: why not use multi-line comments?

File: react/features/base/participants/middleware.js [hunk 7]
```diff
@@ -63,20 +63,22 @@ MiddlewareRegistry.register(store => next => action => {
...
+                conference,
                 id,
                 local: participant.id === id,
                 raisedHand: false
             }));
         }
 
-        typeof APP === 'object'
-            && APP.UI.markDominantSpeaker(action.participant.id);
+        if (typeof APP === 'object') {
```
Comment (saghul) [inline react/features/base/participants/middleware.js]: you always suggest to use the && syntax, why the change here?

File: react/features/base/conference/actions.js [hunk 8]
```diff
@@ -125,13 +125,14 @@ function _addConferenceListeners(conference, dispatch) {
     conference.on(
         JitsiConferenceEvents.DISPLAY_NAME_CHANGED,
         (id, displayName) => dispatch(participantUpdated({
+            conference,
```
Comment (saghul) [inline react/features/base/conference/actions.js]: (I'm putting this comment here, but it applies to this entire commit).

Why do we need to pass the conference when updating a participant? Are we worried that the participant ID might collide? So, assuming that:

- participant IDs cannot collide
- a participant will always be added before it's updated

I don't see why we need the conference for PARTICIPANT_UPDATED. We can just get the participant by ID and the conference will be there. And it's not like that value can change.

File: react/features/base/redux/StateListenerRegistry.js (hunk 4 above)
Comment (lyubomir) [reply react/features/base/redux/StateListenerRegistry.js]: Because it would require guarding (checking whether it's the first register or not) which is more code that the absolute minimum required.

File: react/features/base/redux/StateListenerRegistry.js (hunk 5 above)
Comment (lyubomir) [reply react/features/base/redux/StateListenerRegistry.js]: Because it's easier. Though I haven't added shallow equality, there is still complete flexibility though:

* If a Selector needs to return a simple value, obviously, `===`/`!==` will suffice.
* If the Selector needs to return a combination of simple values, they should rather compute the next/new `selection` based on the `prevSelection`. For example, with the help of our base/redux functions `set` or `assign`, `prevSelection` can immutably be transformed into new/next `selection`.

Anyway, I'm against `_.isEqual` in particular because it's deep equality which is very different from what redux and react-redux give us.

File: react/features/base/conference/actions.js (hunk 8 above)
Comment (lyubomir) [reply react/features/base/conference/actions.js]: Participant IDs can collide across JitsiConference instances, participant IDs are "unique" only within a JitsiConference instance. That's why the identification of a remote participant needs to be the pair id-conference. Which is a lot bigger task than this PR, this PR reduces the number of places in which collisions can have lasting, visible effects on the app.

**Comment by lyubomir** [comment]
> I this there is one legit bug in there when comparing selectors, but we are not running into it yet.

Please do point me to it. I don't want to hit the bug in the future in some hard to comprehend way.

File: react/features/base/participants/reducer.js [hunk 9]
```diff
@@ -39,6 +39,44 @@ declare var APP: Object;
...
+    case DOMINANT_SPEAKER_CHANGED:
+    case PARTICIPANT_ID_CHANGED:
+    case PARTICIPANT_UPDATED:
+    case PIN_PARTICIPANT:
+        return state.map(p => _participant(p, action));
+
+    case PARTICIPANT_JOINED: {
+        const participant = _participant(/* state */ undefined, action);
+
+        // Give the _participant function a chance to deny the addition of the
```
[the remark this answers is not in the document]
Comment (lyubomir) [reply react/features/base/participants/reducer.js]: This is not an argument against what you asked, just thoughts. I almost returned `undefined` from `_participant` in order to reject a participant defined by `action`. So I thought the source code here exactly was a bit schizophrenic: it's delegating all responsibility to the function `_participant` but is saying "I don't know, care what s**t `action` contains but you must accept it and return me a new participant". Anyway, I'll happily oblige you here.

File: react/features/base/participants/reducer.js (hunk 3 above)
Comment (lyubomir) [reply react/features/base/participants/reducer.js]: OK. Will come in "Associate remote participant w/ JitsiConference (_JOINED)".

File: react/features/base/participants/middleware.js (hunk 6 above)
Comment (lyubomir) [reply react/features/base/participants/middleware.js]: Let me understand this please. The style of comment:

````javascript
/**
 * Blah-blah
 */
````

would signify a jsdoc comment (in my understanding).

But what follows it is a function call, not a function definition. So I'm explaining what the "next line" which is a function call intends to implement and that (in my understanding) is not the above style of comment, it's a plain, non-special type of comment.

We're doing this weird mismatch of comment styles on `MiddlewareRegisry` and `ReducerRegistry` uses where the comment should actually be on the argument of the `register` call which is a function definitions. I presume your comment comes from there. While I don't like the weird mismatch of comment styles, I'll address your comment for the sake of consistency.

Comment (lyubomir) [reply react/features/base/participants/middleware.js]: OK. Will come in "Remove remote participants who are no longer of interest".

File: react/features/base/participants/middleware.js (hunk 7 above)
Comment (lyubomir) [reply react/features/base/participants/middleware.js]: Because I'm insecure... I think I changed it in "Coding style" to &&. Then in a later commit I modified the preceding `if` and what thinking: OK, someone's going to see these sequential `if`s and be like "Why is the first one a if and the second one a &&".

Comment (lyubomir) [reply react/features/base/participants/middleware.js]: OK. Will come in "Associate remote participant w/ JitsiConference (_UPDATED)".

**Comment by lyubomir** [comment]
@saghul, please take a look! I believe I address all your comments (existing at the time I'm writing this, of course) by replying to them and/or making the changes I agree with or am OK applying. I didn't make any modifications to source code other than pointed by your comments and I indicated in which comment the changes were made.

I additionally added some more commit messages. I know it's a bit late after a review. I'm sorry, I was wasted last night.

File: react/features/base/participants/reducer.js [hunk 10]
```diff
@@ -147,31 +130,63 @@ function _participant(state: Object = {}, action) {
...
-    switch (action.type) {
-    case DOMINANT_SPEAKER_CHANGED:
-    case PARTICIPANT_ID_CHANGED:
-    case PARTICIPANT_UPDATED:
-    case PIN_PARTICIPANT:
-        return state.map(p => _participant(p, action));
-
-    case PARTICIPANT_JOINED:
-        return [ ...state, _participant(undefined, action) ];
+function _participantJoined(state, { participant }) {
```
Comment (saghul) [inline react/features/base/participants/reducer.js]: state is unused, let's get rid of it?

File: react/features/base/conference/actions.js (hunk 8 above)
Comment (saghul) [reply react/features/base/conference/actions.js]: 👍

File: react/features/base/redux/StateListenerRegistry.js (hunk 5 above)
Comment (saghul) [reply react/features/base/redux/StateListenerRegistry.js]: Ack.

File: react/features/base/redux/StateListenerRegistry.js (hunk 4 above)
Comment (saghul) [reply react/features/base/redux/StateListenerRegistry.js]: FTR, this requires that registration happens prior to the subscription, but I think that's a reasonable expectation, so I'm ok with the outcome here. 👍

File: react/features/base/participants/reducer.js (hunk 10 above)
Comment (lyubomir) [reply react/features/base/participants/reducer.js]: OK. Will come in "Associate remote participant w/ JitsiConference (_JOINED)".

File: react/features/base/redux/StateListenerRegistry.js (hunk 4 above)
Comment (lyubomir) [reply react/features/base/redux/StateListenerRegistry.js]: Correct. That's the common requirement for all "XXXRegistry" like `MiddlewareRegistry` and `ReducerRegistry`.

File: react/features/mobile/image-cache/middleware.js [hunk 11]
```diff
@@ -37,8 +37,13 @@ const _PREFETCH_AVATAR_URLS = false;
 MiddlewareRegistry.register(({ getState }) => next => action => {
     switch (action.type) {
     case APP_WILL_MOUNT:
-    case CONFERENCE_FAILED:
-    case CONFERENCE_LEFT:
+    case SET_LOCATION_URL:
```
Comment (saghul) [inline react/features/mobile/image-cache/middleware.js]: I just realized how often we're doing this. Why not do it only in APP_WILL_MOUNT? Would it be unrealistic to expect the app to eventually close, before it becomes a problem? As it stands, we are potentially downloading avatars over and over again...

File: react/features/base/participants/middleware.js (hunk 6 above)
Comment (saghul) [reply react/features/base/participants/middleware.js]: Sorry I missed this. Generally, I prefer multi-line (/* ... */) comments for multi-line comments. But I know we are not 100% consistent here either.

File: react/features/base/participants/middleware.js (hunk 7 above)
Comment (saghul) [reply react/features/base/participants/middleware.js]: 🤣

File: react/features/mobile/image-cache/middleware.js (hunk 11 above)
Comment (lyubomir) [reply react/features/mobile/image-cache/middleware.js]: OK. Will come in "[RN] Fix a "TypeError: Cannot read property handlers of undefined" in ImageCache".

**Comment by lyubomir** [comment]
> Jenkins, please test this please.

It's testing, it's just that there were may PRs in the queue.

**Comment by saghul** [comment]
Yup, at any rate, I approved it so go ahead and merge once the tests are 💚
